File: backend/app/services/findings_service.py
from __future__ import annotations

import logging
from typing import Any

from pydantic import ValidationError

from app.ai.detector import findings_detector
from app.ai.heuristics import SAMPLE_FINDINGS, HeuristicFinding
from app.ai.llm_service import LLMServiceError, llm_service
from app.db.supabase import DatabaseConfigError, DatabaseQueryError
from app.schemas.ai_findings import AIFindingCreate
from app.services.database import database_service

logger = logging.getLogger(__name__)


class FindingsService:
    """Runs detection, enriches findings, and persists results."""

    def list_findings(self, limit: int = 100) -> dict[str, Any]:
        try:
            return {"findings": database_service.list_findings(limit=limit), "sample_findings": SAMPLE_FINDINGS}
        except (DatabaseConfigError, DatabaseQueryError) as exc:
            logger.info("list_findings_fallback reason=%s", exc)
            return {"findings": [], "sample_findings": SAMPLE_FINDINGS, "errors": [str(exc)]}

    def get_finding(self, finding_id: str) -> dict[str, Any] | None:
        return database_service.get_ai_finding(finding_id)

    async def run_analysis(
        self,
        *,
        integration_id: str | None = None,
        limit: int = 500,
        enrich: bool = True,
    ) -> dict[str, Any]:
        try:
            spend_records = (
                database_service.list_integration_spend(integration_id)
                if integration_id
                else database_service.list_spend_records(limit=limit)
            )
        except (DatabaseConfigError, DatabaseQueryError) as exc:
            logger.exception("findings_spend_load_failed integration_id=%s", integration_id)
            return {
                "detected_count": 0,
                "stored_count": 0,
                "findings": [],
                "sample_findings": SAMPLE_FINDINGS,
                "errors": [str(exc)],
            }

        detected = findings_detector.detect(spend_records)
        stored: list[dict[str, Any]] = []
        skipped_duplicates = 0
        errors: list[str] = []

        for finding in detected:
            try:
                if self._is_duplicate(finding):
                    skipped_duplicates += 1
                    continue
                enriched = await self._enrich(finding) if enrich else {}
                stored.append(self._persist_finding(finding, enriched))
            except (DatabaseConfigError, DatabaseQueryError, ValidationError, LLMServiceError) as exc:
                logger.warning(
                    "finding_store_or_enrich_failed type=%s resource_id=%s",
                    finding.finding_type,
                    finding.resource_id,
                )
                errors.append(f"{finding.finding_type}:{finding.resource_id}: {exc}")

        return {
            "detected_count": len(detected),
            "stored_count": len(stored),
            "skipped_duplicates": skipped_duplicates,
            "findings": stored,
            "errors": errors,
        }

    @staticmethod
    def _is_duplicate(finding: HeuristicFinding) -> bool:
        existing = database_service.find_existing_finding(finding.resource_id, finding.finding_type)
        return existing is not None

    async def _enrich(self, finding: HeuristicFinding) -> dict[str, str]:
        payload = finding.model_dump(mode="json")
        try:
            return await llm_service.enrich_finding(payload)
        except LLMServiceError:
            raise
        except Exception as exc:
            raise LLMServiceError("Unexpected LLM enrichment failure") from exc

    @staticmethod
    def _persist_finding(finding: HeuristicFinding, enrichment: dict[str, str]) -> dict[str, Any]:
        payload = AIFindingCreate(
            resource_id=finding.resource_id,
            finding_type=finding.finding_type,
            severity=finding.severity,
            confidence=finding.confidence,
            estimated_savings=finding.estimated_savings,
            recommended_action=finding.recommended_action,
            reason=finding.reason,
            explanation=enrichment.get("explanation"),
            risk_summary=enrichment.get("risk_summary"),
            optimization_recommendation=enrichment.get("optimization_recommendation"),
        )
        return database_service.create_ai_finding(payload)


findings_service = FindingsService()

File: backend-java/src/main/java/com/killswitch/service/FindingsService.java
package com.killswitch.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.killswitch.ai.GeminiService;
import com.killswitch.ai.HeuristicsEngine;
import com.killswitch.model.Finding;
import com.killswitch.model.SpendRecord;
import com.killswitch.rules.RuleEngine;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class FindingsService {
    private final DatabaseService database;
    private final HeuristicsEngine heuristicsEngine;
    private final GeminiService geminiService;
    private final RuleEngine ruleEngine;
    private final ObjectMapper objectMapper;

    public Map<String, Object> listFindingsEnvelope(int limit, String status, String severity) {
        try {
            Map<String, Object> filters = new LinkedHashMap<>();
            if (status != null && !status.isBlank()) {
                filters.put("status", status);
            }
            if (severity != null && !severity.isBlank()) {
                filters.put("severity", severity);
            }
            return Map.of("findings", database.listFindings(limit, filters), "sample_findings", sampleFindings());
        } catch (Exception e) {
            return Map.of("findings", List.of(), "sample_findings", sampleFindings(), "errors", List.of(e.getMessage()));
        }
    }

    public List<Map<String, Object>> getFindings(int limit, String status, String severity) {
        return castList(listFindingsEnvelope(limit, status, severity).get("findings"));
    }

    public Map<String, Object> getFindingById(String id) {
        return database.getFinding(id);
    }

    public Map<String, Object> runAnalysis(String integrationId, int limit, boolean enrich) {
        List<Map<String, Object>> spendRows;
        try {
            spendRows = integrationId == null || integrationId.isBlank()
                ? database.listSpendRecords(limit)
                : database.listIntegrationSpend(integrationId);
        } catch (Exception e) {
            return Map.of(
                "detected_count", 0,
                "stored_count", 0,
                "findings", List.of(),
                "sample_findings", sampleFindings(),
                "errors", List.of(e.getMessage())
            );
        }

        List<SpendRecord> spendRecords = spendRows.stream()
            .map(row -> objectMapper.convertValue(row, SpendRecord.class))
            .toList();
        List<Finding> detected = heuristicsEngine.analyze(spendRecords);
        List<Map<String, Object>> stored = new ArrayList<>();
        List<String> errors = new ArrayList<>();
        int skipped = 0;

        for (Finding finding : detected) {
            Map<String, Object> payload = findingPayload(finding);
            try {
                if (database.findExistingFinding(String.valueOf(payload.get("resource_id")), String.valueOf(payload.get("finding_type"))) != null) {
                    skipped++;
                    continue;
                }
                if (enrich) {
                    payload.putAll(geminiService.enrichFinding(payload));
                }
                Map<String, Object> created = database.createFinding(payload);
                if (created != null) {
                    stored.add(created);
                }
            } catch (Exception e) {
                errors.add(payload.get("finding_type") + ":" + payload.get("resource_id") + ": " + e.getMessage());
            }
        }

        return Map.of(
            "detected_count", detected.size(),
            "stored_count", stored.size(),
            "skipped_duplicates", skipped,
            "findings", stored,
            "errors", errors
        );
    }

    public Map<String, Object> runAnalysis(boolean enrich) {
        return runAnalysis(null, 500, enrich);
    }

    public Map<String, Object> updateStatus(String id, String status) {
        if (status == null || status.isBlank()) {
            throw new IllegalArgumentException("status is required");
        }
        return database.updateFinding(id, Map.of("status", status));
    }

    public void deleteFinding(String id) {
        database.deleteFinding(id);
    }

    public Map<String, Object> applyRules() {
        List<Map<String, Object>> findings = database.listFindings(500, Map.of("status", "open"));
        List<Map<String, Object>> rules = database.listRules(500);
        List<RuleEngine.Decision> decisions = ruleEngine.evaluate(findings, rules);
        List<Map<String, Object>> actions = new ArrayList<>();
        for (RuleEngine.Decision decision : decisions) {
            actions.add(database.createAction(Map.of(
                "finding_id", decision.finding().get("id"),
                "action_type", decision.actionType(),
                "status", "pending",
                "execution_mode", decision.executionMode(),
                "savings", decision.finding().getOrDefault("estimated_savings", BigDecimal.ZERO),
                "execution_log", "Rule matched; action queued.",
                "rollback_status", "not_required"
            )));
        }
        return Map.of("decisions", decisions.stream().map(RuleEngine.Decision::toMap).toList(), "actions", actions);
    }

    private Map<String, Object> findingPayload(Finding finding) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("resource_id", finding.getResourceId());
        payload.put("finding_type", finding.getFindingType());
        payload.put("severity", finding.getSeverity());
        payload.put("confidence", finding.getConfidence());
        payload.put("estimated_savings", finding.getEstimatedSavings());
        payload.put("recommended_action", finding.getRecommendedAction());
        payload.put("reason", finding.getReason());
        payload.put("explanation", finding.getExplanation());
        payload.put("risk_summary", finding.getRiskSummary());
        payload.put("optimization_recommendation", finding.getOptimizationRecommendation());
        payload.put("status", finding.getStatus() == null ? "open" : finding.getStatus());
        payload.put("provider", finding.getProvider());
        return payload;
    }

    private List<Map<String, Object>> sampleFindings() {
        return List.of(Map.of(
            "id", "sample-zombie-seat",
            "resource_id", "sample-github-seat",
            "finding_type", "unused_seats",
            "severity", "high",
            "confidence", 0.9,
            "estimated_savings", 21,
            "recommended_action", "remove_seat",
            "reason", "Sample GitHub seat inactive for 60+ days.",
            "status", "open",
            "created_at", OffsetDateTime.now().toString()
        ));
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> castList(Object value) {
        return value instanceof List<?> list ? (List<Map<String, Object>>) list : List.of();
    }
}
